core/trino-main/src/test/java/io/trino/operator/TestDeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,443 @@
+        assertThatThrownBy(() -> testPollPages(
+                ImmutableListMultimap.of(
+                        createTaskId(0, 0),
+                        createPage("p0a0v0"),
+                        createTaskId(1, 0),
+                        createPage("p1a0v0"),
+                        createTaskId(0, 1),
+                        createPage("p0a1v0")),
+                ImmutableMap.of(
+                        createTaskId(2, 2),
+                        error),
+                ImmutableList.of("p0a1v0"))).isEqualTo(error);
+        assertThatThrownBy(() -> testPollPages(
+                ImmutableListMultimap.of(
+                        createTaskId(0, 0),
+                        createPage("p0a0v0"),
+                        createTaskId(1, 0),
+                        createPage("p1a0v0"),
+                        createTaskId(0, 1),
+                        createPage("p0a1v0")),
+                ImmutableMap.of(
+                        createTaskId(0, 1),
+                        error),
+                ImmutableList.of("p0a1v0"))).isEqualTo(error);
+    }
+
+    private void testPollPages(Multimap<TaskId, SerializedPage> pages, Map<TaskId, RuntimeException> failures, List<String> expectedValues)
+    {
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            for (TaskId taskId : Sets.union(pages.keySet(), failures.keySet())) {
+                buffer.addTask(taskId);
+            }
+            for (Map.Entry<TaskId, SerializedPage> page : pages.entries()) {
+                buffer.addPages(page.getKey(), ImmutableList.of(page.getValue()));
+            }
+            for (Map.Entry<TaskId, RuntimeException> failure : failures.entrySet()) {
+                buffer.taskFailed(failure.getKey(), failure.getValue());
+            }
+            for (TaskId taskId : Sets.difference(pages.keySet(), failures.keySet())) {
+                buffer.taskFinished(taskId);
+            }
+            buffer.noMoreTasks();
+            assertContainsPages(buffer, expectedValues);
+        }
+    }
+
+    private static void assertContainsPages(ExchangeClientBuffer buffer, List<String> values)
+    {
+        List<String> actual = new ArrayList<>();
+        while (true) {
+            SerializedPage page = buffer.pollPage();
+            if (page == null) {
+                break;
+            }
+            actual.add(page.getSlice().toStringUtf8());
+        }
+        assertTrue(buffer.isFinished());
+        assertThat(actual).containsExactlyInAnyOrderElementsOf(values);
+    }
+
+    @Test
+    public void testRemovePagesForPreviousAttempts()
+    {
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            assertEquals(buffer.getRetainedSizeInBytes(), 0);
+
+            TaskId task1 = createTaskId(0, 0);
```

Comment:
nit: name variables task_0_0, task_1_0 and task_0_1

Reply:
Do we use snake case? Renamed to `task0Attempt0` / `task0Attempt1` for now